text/0000-feature-metadata.md

```diff
@@ -0,0 +1,275 @@
+just a dependency array. This sample section of `Cargo.toml` shows new
+possibilities:
+
+```toml
+[features]
+# Current configuration will continue to work
+foo = []
+# New configurations
+bar = { requires = ["foo"], doc = "simple docstring here"}
+baz = { requires = ["foo"], public = false}
+qux = { requires = [], deprecated = true, unstable = true }
+quux = { requires = [], deprecated = { since = "1.2.3", note = "don't use this!" } }
+
+# Features can also be full tables if descriptions are longer
+[features.corge]
+requires = ["bar", "baz"]
+doc = """
+# corge
+
+This could be a longer description of this feature
+"""
+```
+
+The following keys would be allowed in a feature table:
+
+-   `requires`: This is synonymous with the existing array describing required
+    features. For example, `foo = ["dep:serde", "otherfeat"]` will be identical to
+    `foo = { requires = ["dep:serde", "otherfeat"] }`
+-   `doc`: A markdown docstring describing the feature. Like with `#[doc(...)]`,
+    the first line will be treated as a summary.
+-   `deprecated`: This can be either a simple boolean, a string, or an object
+    with `since` and/or `note` keys. Cargo will warn downstream crates using this
+-   `public`: A boolean flag defaulting to `true` that indicates whether or not
+    downstream crates should be allowed to use this feature
+-   `unstable`: A boolean flag indicating that the feature is only usable with
+    nightly Rust, or is otherwise not API-stable
+
+If a downstream crate attempts to use the features `baz` and `qux`, they will
+see messages like the following:
+
+```
+warning: feature `quux` on crate `mycrate` is deprecated since version
+  1.2.3: "don't use this!"
+
+error: feature `baz` on crate `mycrate` is private and cannot be used by
+  downstream crates
+```
+
+# Reference-level explanation
+
+[reference-level-explanation]: #reference-level-explanation
+
+Validation and parsing of the new schema, described above, shoud be relatively
+straightforward. Each of the added keys is discussed individually in the
+following sections:
+
+## `requires`
+
+`requires` will take the place of the feature dependency array that currently
+exists. Semantics will remain unchanged.
+
+## `doc`
+
+`doc` is the most straightforward: it accepts markdown-flavored text, and should
+be thought of as the equivalent to a `#[doc(...)]` attribute. Like doc comments,
+the first line should be usable as a summary, though most use cases will only
+have one line.
+
+There is nothing in this RFC that cargo `must` do with this action, since it is
+mainly intended for the consumption of `rustdoc` or `docs.rs`. However, it can
+be used for general diagnostic information.
+
+## `deprecated`
+
+`deprecated` should be thought of as the equivalent of the [`deprecated`]
+attribute in Rust source. The value can be a boolean, string, or an object with
+`since` or `note` keys. Schema rules are as follows:
```

Comment:
I suspect this will be needed during resolving which means this will be needed in the Index.  We should verify that and specify any Index-related changes from this or other features

Reply:
I don't think I know enough about this section or how to specify the outcome. Would the behavior change be that Cargo prefers version combinations that don't use deprecated flags?

Reply:
Not that it will change resolution but that most likely the resolver is where we will know the appropriate information to report this.  Unsure though.

Reply:
I'll poke around on Zulip and see if I can clarify that. Added this as a unresolved question

Reply:
In thinking about this, we might be able to add `description` to the index without much problem, if needed.  The big problem is `public` (and `unstable` if the RFC includes it').  The problem here is that `public` restricts how people can use the features.  If we don't include `public` in the index or if we allow old versions of cargo to use entries with `public = false`, then people could use these features and they'd break upon upgrading cargo.  I suspect we wouldn't want that and we should prevent old cargo's from reading these index entries.

However, revving the Index format for this would run into rust-lang/cargo#10623 which we should probably take care of long before this RFC gets implemented.

Reply:
I think this has been resolved since index changes are now mentioned.